### ISSUE TITLE
K6 should add HTTP compression request header by default

### ISSUE THREAD

**Issue by testn** [grafana/k6#2586]
### Feature Description

Currently, k6 http does not add compression header by default. This makes a unpleasant surprise when you find that K6 can load test with lower request rate due to the network bandwidth being full

**Comment by na--** [comment]
Thanks for opening this issue. Unfortunately, we can't make this change in the current `k6/http` API, since it will be a pretty major breaking change. Existing load tests will start returning different measurements with the new k6 version, potentially breaking existing thresholds... Besides, some HTTP servers don't actually support compression, so I am not sure we should ever enable it by default :thinking: 

Still, you raise a very valid point. It should be easy for a k6 user to specify globally that they want compression enabled on all HTTP requests. The current inability to do this in k6 is poor UX and we should fix it.

In the mid-to-long term, this will be addressed by the new HTTP API (https://github.com/grafana/k6/issues/2461) we plan to build for k6, which will probably have a concept of a `Client` and/or `Transport` objects that can accept such properties and apply them to all of the HTTP requests that pass through them. In the short term, I'd advise to modify the [`httpx` JS wrapper](https://k6.io/docs/javascript-api/jslib/httpx/) around `k6/http` we have hosted on [jslib.k6.io](https://jslib.k6.io/) to automatically  add compression to all of the individual requests.

**Comment by testn** [comment]
I think this is quite unexpecting for most of users. When we bump a major version, we should start doing this by default but provide an option to turn it off.

**Comment by mstoykov** [comment]
I did look around and as far as I can see none of the fetch API, python's requests or golang stdlib automatically compress the body of a request.

So I am kind of wondering why would you expect that k6 will be doing it?

**Comment by testn** [comment]
Not the body of the request. I'm talking about sending "Accept-Encoding: br, deflate, gzip" to the server to indicate that it can handle compressed response

**Comment by na--** [comment]
:man_facepalming: both me and @mstoykov thought you were talking about the `compression` option in https://k6.io/docs/javascript-api/k6-http/params/ :sweat_smile:

**Comment by na--** [comment]
Yeah, to get back on topic, adding `Accept-Encoding` by default will be much less of a problem than `Content-Encoding`... :thinking: And a lot of HTTP clients (browsers especially) add it by default, so it makes a lot of sense for k6 to do it too. But it will still kind of be a breaking change, as you say there is a difference in performance, so already existing thresholds will be affected without any change in the scripts or the systems under test themselves, just by updating the used k6 version... :thinking: 

So while I agree this should have probably been the sensible default in the current HTTP API, if we were creating it today, and it should be the default in the upcoming new k6 HTTP API (https://github.com/grafana/k6/issues/2461), I am on the fence if we should change the current HTTP API with it :confused: On the one hand, anyone that wants it can just use the [`addHeader()` method of the `httpx` wrapper](https://k6.io/docs/javascript-api/jslib/httpx/addheader) around it, without having to modify each request. On the other hand, they first would have to realize k6 doesn't automatically send it, and they may need to rewrite their tests to use httpx... So yeah, I am on the fence. @grafana/k6-devs, WDYT?

**Comment by testn** [comment]
> From a quick test k6 does add `Accept-Encoding` with a value of `gzip`

Hmm... I must have missed something. I tried it in 0.33 and it didn't add ```Accept-Encoding``` header by default.

**Comment by mstoykov** [comment]
apperantly I was fooled by https://github.com/grafana/k6/issues/1042 , and having now checked with the keylogger from #2487 - we definitely do not send `Accept-Encoding` by default :(

**Comment by oleiade** [comment]
My partial understanding leads me to expect affected thresholds would be mostly impacted positively? In the sense that because we would automatically advertising to the server that k6 now accepts more efficient (compressed) encodings, in more scenarios, the performance would get better. Most thresholds I've seen so far in the form of `metric value < target value`. Thus, if `metric value` goes down, it might not be such an issue after all? 🤷🏻 

On the other hand, I read here and there that the recommendation is to not using `Accept-Encoding` against servers with high loads (Microsoft argues >80% CPU load). 

My rather conservative position on this would be to argue that no one noticed or complained about it until, thus, it might not be a top priority. On the other hand, from a more UX perspective, I can see how k6 should be a good citizen, and use sensible defaults, and I believe using `Accept-Encoding` is a sensible default.

Thus, because we have this milestone coming of refactoring the HTTP api, I believe this is the perfect opportunity to make that change. Anything else feels like it would need a lot of extensive communication with our users, and in our documentation, without the insurance that we're not breaking things silently for our users.

**Comment by testn** [comment]
If you pass in ```Accept-Encoding```, the server does not actually have to respond back to you with compressed content. However, if it chooses to do so, it can pass ```Content-Encoding``` with ```gzip/br/deflate``` back to indicate the content is compresed in the given format.  If not, it can just passed the data as is without any compression.

**Comment by testn** [comment]
I'm not suggesting that you must do it now. I am just trying to convince you that it should be the default behavior. It should be okay to do so if you are thinking of releasing a major version where you can introduce a breaking change.

If you are not convinced, you can also either

1. introduce the change by default but have an option for users to turn it back to the original behavior
2. not introduce the change by default but provide an option for users to ask for the compressed content globally.

**Comment by oleiade** [comment]
Hey @testn 

Thanks a lot for your input! I've learned something, I didn't consider that `Accept-Encoding` was indeed a best effort oriented contract between the client and server. Glad you clarified that  🙇🏻

**Comment by jberryman** [comment]
> My partial understanding leads me to expect affected thresholds would be mostly impacted positively (as in: it wouldn't lead them to be newly crossed as result)? 

@oleiade In hasura the overhead of compression is significant across all the queries we benchmark, so I wouldn't expect this to be true, at least when skipping the network

Even as someone who failed to set `Accept-Encoding: gzip` in their benchmarks until recently, and acknowledging that yes most real http clients will set this header,  i'm opposed to k6 setting headers automatically. It's just as likely that a user would find the proposed Behavior surprising (e.g. If using k6 to profile application code, one might wish and expect to be bypassing the gzip code path, which one has no control over for the most part)

Instead can I please suggest adding a warning for not including `Accept-Encoding`, and other potential foot guns

**Comment by oleiade** [comment]
@jberryman Thanks a lot for your input and insights, those are valuables indeed 🙇🏻 much appreciated!

**Comment by testn** [comment]
I would have a different opinion.... Not gzipping by default seems to be more of a surprise especially on an app where you can saturate the network easily.

If we somehow decide not to implement this as a default, we should at the very least have the option to add `Accept-Encoding: gzip` globally.

**Comment by claytonneal** [comment]
hiya, would prefer that the default is no such header
We often want to benchmark our software with and without gzip - as we provide API's to customers, and its their choice of what to use. So if the default is no such header, and via config it can be turned on that would be great.
